BaseRecalibratorDataflowIntegrationTest on SparkPipelineRunner

tomwhite opened the pull request:
This gets BaseRecalibratorDataflowIntegrationTest running on Spark.

I noticed that the test with "-knownSites" from BaseRecalibratorIntegrationTest (i.e. the non-dataflow version) fails with both the Direct and Spark runners. I had a look at the file output and there are a few discrepancies (see diff below). Any thoughts on how to fix this? I've commented out the test in this PR.

```
diff /var/folders/d1/8f5_j4hx04z72w6wgqxkb2l40000gn/T/walktest.tmp_param.02172067147450353519.tmp src/test/resources/org/broadinstitute/hellbender/tools/BQSR/expected.NA12878.chr17_69k_70k.2inputs.txt
60c60
<           34   3051              34
---
>           34   3050              34
71c71
<           45  46942              45
---
>           45  46940              45
124,126c124,126
< 809R9ABXX101220.5  D                   45.0000             45.0000         23471    0.00
< 809R9ABXX101220.5  I                   45.0000             45.0000         23471    0.00
< 809R9ABXX101220.5  M                   27.0000             27.0494         23471   49.13
---
> 809R9ABXX101220.5  D                   45.0000             45.0000         23470    0.00
> 809R9ABXX101220.5  I                   45.0000             45.0000         23470    0.00
> 809R9ABXX101220.5  M                   27.0000             27.0493         23470   49.13
155c155
< 809R9ABXX101220.5            34  M                   34.0000          3051    2.96
---
> 809R9ABXX101220.5            34  M                   34.0000          3050    2.96
161,162c161,162
< 809R9ABXX101220.5            45  D                   45.0000         23471    0.00
< 809R9ABXX101220.5            45  I                   45.0000         23471    0.00
---
> 809R9ABXX101220.5            45  D                   45.0000         23470    0.00
> 809R9ABXX101220.5            45  I                   45.0000         23470    0.00
2714c2714
< 809R9ABXX101220.5            34  29              Cycle          M                   34.0000            20    0.00
---
> 809R9ABXX101220.5            34  29              Cycle          M                   34.0000            19    0.00
2773c2773
< 809R9ABXX101220.5            34  CA              Context        M                   34.0000           506    0.00
---
> 809R9ABXX101220.5            34  CA              Context        M                   34.0000           505    0.00
3464,3465c3464,3465
< 809R9ABXX101220.5            45  29              Cycle          D                   45.0000           180    0.00
< 809R9ABXX101220.5            45  29              Cycle          I                   45.0000           180    0.00
---
> 809R9ABXX101220.5            45  29              Cycle          D                   45.0000           179    0.00
> 809R9ABXX101220.5            45  29              Cycle          I                   45.0000           179    0.00
3634,3635c3634,3635
< 809R9ABXX101220.5            45  GCA             Context        D                   45.0000           278    0.00
< 809R9ABXX101220.5            45  GCA             Context        I                   45.0000           278    0.00
---
> 809R9ABXX101220.5            45  GCA             Context        D                   45.0000           277    0.00
> 809R9ABXX101220.5            45  GCA             Context        I                   45.0000           277    0.00
```

tomwhite commented:
@droazen or @lbergelson - would you be able to take a look at this one please?

lbergelson commented:
Sorry for the wait tom.  We're a bit behind on pull requests.  We will get to this as soon as possible.

tomwhite commented:
Thanks @lbergelson. I've rebased on master to fix conflicts with the patch you did to remove secret keys. In doing so I changed back your change to BaseRecalibratorDataflowIntegrationTest which added the API key in one place since one of its tests now runs locally again, and there is no API key in that case.

jean-philippe-martin commented:
@tomwhite, the local BaseRecalibratorDataflow needs the API key because, even though it's running locally, it fetches reference bases from the Google Genomics API. 

That is also why I put it in the "cloud" group: my understanding is that this is the group of tests that will fail if you don't have valid credentials for the Google cloud.

tomwhite commented:
@jean-philippe-martin, this change changes `testBQSRLocal` to be truly local, so that it reads the reference from the local filesystem (which was added in https://github.com/broadinstitute/hellbender/pull/827). So, for that test no API key is needed. 

The `testBQSRRefCloud` test still loads the reference from the cloud, so we're still testing that case too.

jean-philippe-martin commented:
@tomwhite ah ok, then that's fine.

tomwhite commented:
I should have added this test in #827, but I missed that all the tests were running in the cloud.

lbergelson commented:
Hmn, commented directly on one of the commits, sorry about that.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/dataflow/pipelines/BaseRecalibratorDataflowIntegrationTest.java:
```diff
@@ -102,14 +106,32 @@ public void runAllTests() throws Exception {
+        return new Object[][]{
+            {new BQSRTest(hg18Reference, HiSeqBam, dbSNPb37, "", getResourceDir() + "expected.NA12878.chr17_69k_70k.txt")},
+            // TODO: fix this (expected and actual output differ in a few places)
+            //{new BQSRTest(hg18Reference, HiSeqBam, dbSNPb37, "-knownSites " + moreSites, getResourceDir() + "expected.NA12878.chr17_69k_70k.2inputs.txt")},
```
I created a new ticket for fixing this (#883), could you add that number into the comment so people know what it's relevant too.

lbergelson commented:
@tomwhite Back to you.  Feel free to squash and merge once it's no longer outputting the api key into the logs.

akiezun commented:
@lbergelson can you take this? you know better than i what to do

lbergelson commented:
This is still leaking api keys into the public logs.  See https://storage.googleapis.com/hellbender/test/build_reports/2311.1/tests/classes/org.broadinstitute.hellbender.tools.dataflow.pipelines.BaseRecalibratorDataflowIntegrationTest.html

akiezun commented:
@lbergelson should we migrate this to hellbender-dataflow?

tomwhite commented:
The non-test changes from this one were committed via https://github.com/broadinstitute/hellbender/pull/924, so the only thing needed is  the test to use a local reference file.

lbergelson commented:
Closing this since it's no longer relevant.  Pushed the branch to gatk-dataflow
